### PR TITLE
Generic the capability to enrich JWT and ID_token with custom claims

### PR DESCRIPTION
**Pull request by Sachin-Mamoru**
#### Purpose:
Introduces a constant to identify dynamic token data functions, enhancing the security and customization of token handling.

#### Approach:
- New Constant Introduction: A constant DYNAMIC_TOKEN_DATA_FUNCTION is added to identify scripts related to token data dynamically, improving maintainability and readability.
- Code Simplification: The condition checking whether user consent is available before fetching token attributes is removed, thereby simplifying the flow and always ensuring extended token attributes are considered.
- Improved Token Attribute Handling: Modifies the condition to check for the presence of a dynamic token data function directly in the script content, ensuring more accurate and secure token handling.

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 1]
```diff
@@ -1931,8 +1930,8 @@ private AccessTokenExtendedAttributes getExtendedTokenAttributes(OAuthMessage oA
         try {
             ServiceProvider serviceProvider = getServiceProvider(oauth2Params.getClientId());
             // TODO: Improve to read the script separately instead of reading from adaptive script.
-            if (!EndpointUtil.isExternalConsentPageEnabledForSP(serviceProvider) ||
-                    serviceProvider.getLocalAndOutBoundAuthenticationConfig().getAuthenticationScriptConfig() == null) {
+            if (!serviceProvider.getLocalAndOutBoundAuthenticationConfig().getAuthenticationScriptConfig()
```
Comment (Thumimku) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:1933]: NPE will occur with there is no authentication script config?

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 2]
```diff
@@ -1931,8 +1930,8 @@ private AccessTokenExtendedAttributes getExtendedTokenAttributes(OAuthMessage oA
...
+                    .getContent().contains(DYNAMIC_TOKEN_DATA_FUNCTION)) {
```
Comment (Thumimku) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:1934]: here also NPE can occur?

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 1 above)
Comment (Sachin-Mamoru) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: Fixed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/2437/commits/007afb081cec57305d9c7387bb5599d49b9b148e

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 2 above)
Comment (Sachin-Mamoru) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: Fixed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/2437/commits/007afb081cec57305d9c7387bb5599d49b9b148e

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 3]
```diff
@@ -1918,9 +1920,7 @@ private void setAuthorizationCode(OAuthMessage oAuthMessage, OAuth2AuthorizeResp
         authorizationResponseDTO.getSuccessResponseDTO().setAuthorizationCode(authorizationCode);
 
         AccessTokenExtendedAttributes tokenExtendedAttributes = null;
-        if (isConsentResponseFromUser(oAuthMessage)) {
-            tokenExtendedAttributes = getExtendedTokenAttributes(oAuthMessage, oauth2Params);
-        }
+        tokenExtendedAttributes = getExtendedTokenAttributes(oAuthMessage, oauth2Params);
```
Comment (Thumimku) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:1923]: for each authz call service provider instance will be retrieve from FW, this is not ideal fix, But approve for now.

**Review by Thumimku — DISMISSED** [review]

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 3 above)
Comment (malithie) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:1923]: We have to have the check whether the app has this function implemented in script. That's a must. We should not let this impact the 80% of the use cases

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 4]
```diff
@@ -1931,8 +1931,11 @@ private AccessTokenExtendedAttributes getExtendedTokenAttributes(OAuthMessage oA
         try {
             ServiceProvider serviceProvider = getServiceProvider(oauth2Params.getClientId());
             // TODO: Improve to read the script separately instead of reading from adaptive script.
-            if (!EndpointUtil.isExternalConsentPageEnabledForSP(serviceProvider) ||
-                    serviceProvider.getLocalAndOutBoundAuthenticationConfig().getAuthenticationScriptConfig() == null) {
+            AuthenticationScriptConfig scriptConfig = serviceProvider.getLocalAndOutBoundAuthenticationConfig()
+                    .getAuthenticationScriptConfig();
+            boolean isValidScript = scriptConfig != null && scriptConfig.getContent() != null &&
+                    scriptConfig.getContent().contains(DYNAMIC_TOKEN_DATA_FUNCTION);
+            if (!isValidScript) {
```
Comment (malithie) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:1938]: ok I see you have moved this to the invocation point. Can we have a separate function like "implementsTokenAttributeExtension (ServiceProvider sp)" which checks this logic and return true or false and then allows invoking this "getExtendedTokenAttributes" function

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 5]
```diff
@@ -263,6 +264,7 @@ public class OAuth2AuthzEndpoint {
     private static final String SET_COOKIE_HEADER = "Set-Cookie";
     private static final String REGEX_PATTERN = "regexp";
     private static final String OIDC_SESSION_ID = "OIDC_SESSION_ID";
+    public static final String DYNAMIC_TOKEN_DATA_FUNCTION = "dynamicTokenData = function (";
```
Comment (malithie) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:267]: You are using this constant in the same class. So make this private. And given that we need to anyway move out from current behavior we should not introduce public constants like these

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 4 above)
Comment (Sachin-Mamoru) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: Fixed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/2437/commits/6ec88a197a4a6608bbc8a1e69f6c53c085d4f54f

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 5 above)
Comment (Sachin-Mamoru) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: Fixed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/2437/commits/6ec88a197a4a6608bbc8a1e69f6c53c085d4f54f

**Comment by malithie** [comment]
The dynamicTokenData function has been implemented very specific to a particular OB use case, which needs to be generalized more.
Even the effort to generalize that function to somelevel here doesn't seem to be working. So we will look at fixing this properly with the right extension points

**Pull request closed without merging**